plutus-ledger-api/src/Plutus/ApiCommon.hs

```diff
@@ -343,3 +347,52 @@ evaluateScriptCounting lv pv verbose ectx p args = swap $ runWriter @LogOutput $
     tell logs
     liftEither $ first CekError $ void res
     pure final
+
+
+{-| A valid parameter name has to be enumeration, bounded, ordered, and
+prettyprintable in a "lowerKebab" way.
+
+Each API version should expose such an enumeration as an ADT and create
+an instance of ParamName out of it.
+-}
+class (Enum a, Ord a, Bounded a) => IsParamName a where
+   showParamName :: a -> String
+   default showParamName :: (Generic a, GIsParamName (Rep a)) => a -> String
+   showParamName = gshowParamName . from
+
+-- | A datatype-generic class to prettyprint 'sums of nullary constructors' in lower-kebab syntax.
+class GIsParamName f where
+    gshowParamName :: f p -> String
+
+instance (GIsParamName a) => GIsParamName (M1 D i a) where
+    gshowParamName (M1 x) = gshowParamName x
+
+instance Constructor i => GIsParamName (M1 C i U1) where
+    gshowParamName = lowerKebab . conName
+      where
+        lowerKebab :: String -> String
+        lowerKebab (h:t) = toLower h : fmap (\c -> if c == '\''
+                                                  then '-'
+                                                  else c
+                                               ) t
+        lowerKebab _ = error "this should not happen"
+
+instance (GIsParamName a, GIsParamName b) => GIsParamName ((:+:) a b) where
+    gshowParamName (L1 x) = gshowParamName x
+    gshowParamName (R1 x) = gshowParamName x
+
+-- | Return all param names given a specific datatype (enumeration).
+enumParamNames :: IsParamName k => [k]
```

Comment:
This is just https://hackage.haskell.org/package/extra-1.7.10/docs/Data-List-Extra.html#v:enumerate ?